src/Castle.ts

```diff
@@ -0,0 +1,218 @@
+    overrideFetch = fetch,
+    failoverStrategy = 'allow',
+  }: {
+    apiSecret: string;
+    apiUrl?: string;
+    timeout?: number;
+    allowedHeaders?: string[];
+    disallowedHeaders?: string[];
+    overrideFetch?: any;
+    failoverStrategy?: string;
+  }) {
+    if (!apiSecret) {
+      throw new Error(
+        'Castle: Unable to instantiate Castle client, API secret is missing.'
+      );
+    }
+
+    this.apiSecret = apiSecret;
+    this.apiUrl = apiUrl || defaultApiUrl;
+    this.timeout = timeout;
+    this.allowedHeaders = allowedHeaders;
+    this.disallowedHeaders = disallowedHeaders;
+    this.overrideFetch = overrideFetch;
+    this.failoverStrategy = failoverStrategy;
+  }
+
+  public async authenticate(
+    params: TrackParameters
+  ): Promise<AuthenticateResult> {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling authenticate.');
+    }
+
+    let response: Response;
+    const controller = new AbortController();
+    const timeout = setTimeout(() => {
+      controller.abort();
+    }, this.timeout);
+
+    try {
+      response = await this.getFetch()(`${this.apiUrl}/v1/authenticate`, {
+        signal: controller.signal,
+        method: 'POST',
+        headers: this.generateDefaultRequestHeaders(),
+        body: this.generateRequestBody(params),
+      });
+    } catch (e) {
+      if (isTimeout(e)) {
+        return {
+          action: this.failoverStrategy,
+          failover: true,
+          failover_reason: 'timeout',
+          user_id: params.user_id,
+        };
+      } else {
+        throw e;
+      }
+    } finally {
+      clearTimeout(timeout);
+    }
+
+    this.handleUnauthorized(response);
+
+    if (!response.ok) {
+      throw new Error(
+        `Castle: \`/authenticate\` request unsuccessful. Expected ok result, got ${
+          response.status
+        }`
+      );
+    }
+
+    return response.json();
+  }
+
+  public async track(params: TrackParameters) {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling track.');
+    }
+
+    let response: Response;
+    const controller = new AbortController();
+    const timeout = setTimeout(() => {
+      controller.abort();
+    }, this.timeout);
+
+    try {
+      response = await this.getFetch()(`${this.apiUrl}/v1/track`, {
+        signal: controller.signal,
+        method: 'POST',
+        headers: this.generateDefaultRequestHeaders(),
+        body: this.generateRequestBody(params),
+      });
+    } catch (e) {
+      if (isTimeout(e)) {
+        // tslint:disable-next-line:no-console
+        console.error(
+          `Castle: an exception occured while tracking ${
+            params.event
+          } event. Request timed out.`
+        );
+        return;
+      }
+    } finally {
+      clearTimeout(timeout);
+    }
+
+    this.handleUnauthorized(response);
+
+    if (!response.ok) {
+      throw new Error('Castle: `/track` request unsuccessful.');
```

Comment:
is there a chance we could include some sort of debug info when errors are being thrown? This would help potential programmers in their debug as well

Reply:
Hmm, I've yet to add optional request logging. That would cover this use case I think.